### DISCUSSION TITLE
UML Difference - internal transitions

### DISCUSSION THREAD

**Discussion by adamfk** [StateSmith/StateSmith#158] [Ideas]
From this discussion in a ticket: https://github.com/StateSmith/StateSmith/issues/136#issuecomment-1493832586

StateSmith treats non-transition behaviors (called internal transitions in UML) differently than the UML spec.

This is intentional because it is very helpful and flexible. That said, I'm open to changing this to be UML compliant if we can keep the benefits StateSmith currently has.

Let's start with a simple example comparing StateSmith and UML semantics. The desire is to transition from state `S1` to `S2` after 3 `INCREASE` events.

![image](https://user-images.githubusercontent.com/274012/229882408-a87daff2-a254-4ea9-b8f0-605e7e49123c.png)

The StateSmith semantics are not UML compliant because `1. INCREASE / count++;` should consume the event with an "internal transition" and prevent the transition to S2 `2. INCREASE [count >= 3]`. Instead, StateSmith non-transition behaviors ("internal transitions") still allow other state behaviors to be checked. This allows a state to re-use a common setup behavior.

The StateSmith example on the left is my personal favorite. Clear and concise. Is there a better way to achieve that desired behavior in UML than what I've shown above?

--- 

The main problem I see with UML semantics is that it prevents behavior re-use or leads to overly complicated designs.

---

Here's another example. In this communication state machine, S1 needs to do the following when receiving the `PACKET_RX` event:
1. process the packet with `s1_process_packet();`. Must happen before `g2( )` and `g3( )` checks.
2. check if it should transition to S2 based on `g2( )`.
3. check if it should transition to S3 based on `g3( )`.

Compare the StateSmith and UML compliant solutions.

![image](https://user-images.githubusercontent.com/274012/229886760-881575b5-b63a-4e1f-9b27-57cabdef9289.png)

None of them are bad, but I still prefer the StateSmith version. Most performant, easy to use, clear, concise. This pattern actually has come up a lot in some of my designs. It is very handy.

--- 

We've seen the benefits of the StateSmith approach, but how do we achieve the UML equivalent of the below design?

![image](https://user-images.githubusercontent.com/274012/229892721-b2a6a384-fabf-45e5-8c17-4f0c07e255d0.png)

The StateSmith options look pretty decent to me.

Editable draw.io image file if you would like to modify the examples: 
![non-transition-behavior-Page-1 drawio](https://user-images.githubusercontent.com/274012/229892948-a8e14eec-38d1-4585-9f0d-5e129aa84166.svg)

Thoughts?

**Comment by iamjuing** [comment]
Hi Adam,

from your description on Wiki...
_"I'm trying to balance the needs of UML users and regular users. When I write documentation, I try to avoid confusing UML terminology like "internal transitions". Most users that don't come from a UML background find the UML "internal transitions" terminology confusing because there is no actual transition between states. It also gets worse when we add in UML "local transitions" as it sounds very much like an "internal transition"."_

Well, anybody who's getting started with state machines will certainly come across the UML. As far as I know, it is the only standard for state machines so, it is naturally the first point of getting touch with it. The question is: Who you call is the "regular" user? For me it is somebody (most likely a professional) who is aware of the UML standard and uses it regularly.

I understand it is difficult to support advanced concepts (like sub-machines, orthogonality etc.) of the standard because of limitations of generic drawing/modeling tools (I encountered that myself when doing generator for Enterprise Architect where I leveraged OOP stereotypes to give elents more meaning) however; I believe, the tool shall follow the UML standard at the first place, so the tool can be used by professionals and UML "followers" effortlessly. Need to stress that this strategy does not forbid extensions at all.

_"The StateSmith options look pretty decent to me."_
_"None of them are bad, but I still prefer the StateSmith version."_
I recommend to stay away of feelings or preferences here. Once the tool supports the UML standard at a decent level (and it is not far from it), it can start providing "shortcuts" ... which I like and totally support, as I do agree with you that achieving some design goals can lead to complex, unreadable but still correct diagrams.

The key point here is, not to confuse the designers using UML, which I bet, are the vast majority of the (future) users; as I see StateSmith as a very promising tool.

Regarding the internal and local transitions ... there is nothing confusing about them. The opposite is true. They are very well defined. They just provide a way how to react to external events w/o leaving the state. Nothing less, nothing more. They only differ in the target node.

Now, to the examples ...
**LightSM problem in UML:**
![image](https://user-images.githubusercontent.com/125978139/230720096-60e2693e-a856-4a1a-954c-4d88c6df11d4.png)

**Protocol example:**
![image](https://user-images.githubusercontent.com/125978139/230722134-1b51f6cb-3c33-4d58-8ea4-0b789f7c9dcd.png)

As you typed, it can sometimes look overly complicated, but still perfectly clear.

**The bottom examples:**
_"but how do we achieve the UML equivalent of the below design?"_
CountSm1 is the answer, right?
Moreover, it is much clearer and complete than CountSm2 and CountSm3.
I even find CountSm2 incorect as it does not specify what to do if none of those transitions gets enabled. What the state machine must do in that case, is to send the unprocessed event to the parent state which is certainly not intended by the designer.

**Conclusion:**
Recommend to support UML at the first place. The UML is pretty mature and already resolves issues that custom design might introduced. On the other hand, I do not see a reason why the two shall not coexist. UML first, extensions allowed.

**Comment by adamfk** [comment]
Hi Juing,

From my own personal experiences, most people that I know professionally or through university are aware of basic FSM (Mealy/Moore) or maybe MATLAB Stateflow. Most are unaware of UML2 HSMs.

That said, I think you make some convincing points and I'm sure that there are a good number of existing or potential StateSmith users that would prefer UML semantics. I don't want to make a behavior breaking change for the existing StateSmith users lightly though.

What about introducing a new way of defining a state machine if the user wants UML semantics? Something like `$statemachine(uml): MySm` or `$sm(uml): MySm`? This way it will be clear which semantics the state machine follows in the diagram.

Thanks for the diagrams too! I hadn't thought of using a choice pseudo state like that. I'll have to add it to the specifications tests.

You are correct about CountSm2 S1 needing a handler to prevent a parent state from handling the event. I forgot to add that because I didn't draw a parent state.

I have a few more questions, but I'm away from my computer for a few days and can't draw a design.

**Comment by iamjuing** [comment]
Regarding _"I hadn't thought of using a choice pseudo state like that. I'll have to add it to the specifications tests."_
I wasn't sure about that much, so I dug into the UML standard and it is actually forbidden by statement:
**A local Transition can only exist within a composite State.**
So, please ignore that one (could be a handy extention though).

Also, I've noticed `[else]` guard in transitions is also forbidden in UML by statement:
**<guard> is a Boolean expression**
It's most likely an extension that we were using many years ago in our generator and I was just used to it.
Basically, it is just a simplification so there is no need to repeat similar guard in all transitions and just consuming the event no matter what.

So, basically only the two diagrams on the very top are compliant with the UML.

_"I don't want to make a behavior breaking change for the existing StateSmith users lightly though."_
I definitely agree with that.
I was playing a bit with various combinations (SVG below) and found out the recently implemented `[else]` does not work as I originally anticipated anyway (calling `action3()` even when `action1()` is called). I believe it should be even removed as it introduces mess to the current StateSmith behavior I was not aware of.

So introducing `$statemachine(uml): MySm` to get closer to the UML might be the only way to go.
Sounds like too much work though.

![test_fsm drawio](https://user-images.githubusercontent.com/125978139/231193920-3241e9d1-417f-46c3-a712-84234932259b.svg)

**Comment by adamfk** [comment]
I agree that the current `[else]` implementation is misleading. Opened a ticket to either fix or remove. Good call. https://github.com/StateSmith/StateSmith/issues/166

Maybe we explore the UML semantics variant in the future if there is a lot of interest in it and people are willing to help implement. I agree that it would be a fair amount of work.

In the meantime, perhaps we can help achieve some of the UML behavior using this idea: https://github.com/StateSmith/StateSmith/issues/167
